File: cordova/src_plugins/readbeyond-plugin-librarian/src/android/FormatHandlerABZ.java
//
//            _           _            _ 
//           (_)         | |          | |
//  _ __ ___  _ _ __  ___| |_ _ __ ___| |
// | '_ ` _ \| | '_ \/ __| __| '__/ _ \ |
// | | | | | | | | | \__ \ |_| | |  __/ |
// |_| |_| |_|_|_| |_|___/\__|_|  \___|_|
//
// Author:      Alberto Pettarin (www.albertopettarin.it)
// Copyright:   Copyright 2013-2015, ReadBeyond Srl (www.readbeyond.it)
// License:     MIT
// Email:       minstrel@readbeyond.it
// Web:         http://www.readbeyond.it/minstrel/
// Status:      Production
//

package it.readbeyond.minstrel.librarian;

import java.io.File;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.json.JSONObject;

public class FormatHandlerABZ extends FormatHandlerAbstractZIP {

    // ABZ format
    private static final String   ABZ_FORMAT             = "abz";
    private static final String   ABZ_DEFAULT_COVER_JPG  = "cover.jpg";
    private static final String   ABZ_DEFAULT_COVER_PNG  = "cover.png";
    private static final String   ABZ_EXTENSION_PLAYLIST = ".m3u";
    private static final String   ABZ_FILE_METADATA      = "metadata.txt";
    private static final String   ABZ_M3U_HEADER         = "#EXTM3U";
    private static final Pattern  ABZ_M3U_LINE_PATTERN   = Pattern.compile("#EXTINF:([0-9]+),(.*)");
    private static final String   ABZ_M3U_LINE_PREAMBLE  = "#EXTINF:";
    private static final String[] ABZ_AUDIO_EXTENSIONS   = { ".aac",
                                                             ".flac",
                                                             ".m4a",
                                                             ".mp3",
                                                             ".mp4",
                                                             ".oga",
                                                             ".ogg",
                                                             ".wav",
                                                             ".webm" };

    public FormatHandlerABZ() {
        super();
        this.setFormatName(ABZ_FORMAT);
    }

    public Publication parseFile(File file) {
        Publication p = super.parseFile(file);

        Format format = new Format(ABZ_FORMAT);

        try {
            ZipFile zipFile = new ZipFile(file, ZipFile.OPEN_READ);
            Enumeration<? extends ZipEntry> zipFileEntries = zipFile.entries();
            boolean foundMetadataFile   = false;
            boolean foundCover          = false;
            boolean foundPlaylist       = false;
            String zeCover              = null;
            String zePlaylist           = null;
            List<String> assets         = new ArrayList<String>();
            while (zipFileEntries.hasMoreElements()) {
                ZipEntry ze = zipFileEntries.nextElement();
                String name = ze.getName();
                String lower = name.toLowerCase();
                
                if (lower.endsWith(ABZ_DEFAULT_COVER_JPG) || lower.endsWith(ABZ_DEFAULT_COVER_PNG)) {
                    zeCover = name;
                } else if (lower.endsWith(ABZ_EXTENSION_PLAYLIST)) {
                    zePlaylist = name;
                } else if (this.fileHasAllowedExtension(lower, ABZ_AUDIO_EXTENSIONS)) {
                    p.isValid(true);
                    assets.add(name);
                } else if (lower.endsWith(ABZ_FILE_METADATA)) {
                    p.isValid(true);
                    foundMetadataFile = true;
                    if (this.parseMetadataFile(zipFile, ze, format)) {
                        if (format.getMetadatum("internalPathPlaylist") != null) {
                            foundPlaylist = true;
                        }
                        if (format.getMetadatum("internalPathCover") != null) {
                            foundCover = true;
                        }
                    }
                } // end if metadata
            } // end while
            zipFile.close();
           
            // set cover
            if ((! foundCover) && (zeCover != null)) {
                format.addMetadatum("internalPathCover", zeCover);
            }
           
            // default playlist found?
            if ((!foundPlaylist) && (zePlaylist != null)) {
                format.addMetadatum("internalPathPlaylist", zePlaylist);
            }
            
            // set number of assets
            format.addMetadatum("numberOfAssets", "" + assets.size());
            
        } catch (Exception e) {
            // invalidate publication, so it will not be added to library
            p.isValid(false);
        } // end try

        p.addFormat(format);

        // extract cover
        super.extractCover(file, format, p.getID());

        return p;
    }

    public String customAction(String path, JSONObject parameters) {
        if (parameters != null) {
            String command = parameters.optString("command", null);
            if (command != null) {
                
                // get sorted list of assets
                if (command.equals("getSortedListOfAssets")) {
                    String playlistEntry  = parameters.optString("internalPathPlaylist", null);
                    List<ZipAsset> assets = new ArrayList<ZipAsset>();
                    if ((playlistEntry == null) || (playlistEntry.equals(""))) {
                        assets = this.getSortedListOfAssets(path, ABZ_AUDIO_EXTENSIONS);
                    } else {
                        assets = this.parseM3UPlaylistEntry(path, playlistEntry);
                    }
                    return Librarian.stringify(assets, "assets");
                }

            }
        }
        return "";
    }

    private List<ZipAsset> parseM3UPlaylistEntry(String path, String playlistEntry) {
        List<ZipAsset> assets = new ArrayList<ZipAsset>();
        try {
            ZipFile zipFile = new ZipFile(new File(path), ZipFile.OPEN_READ);
            ZipEntry ze     = zipFile.getEntry(playlistEntry);
            String text     = this.getZipEntryText(zipFile, ze);
            String[] lines  = text.split("\n");
            
            // check that the first line starts with the M3U header
            if ((lines.length > 0) && (lines[0].startsWith(ABZ_M3U_HEADER))) {
                for (int i = 1; i < lines.length; i++) {
                    String line = lines[i].trim();
                    
                    // if line starts with the M3U preamble, parse it
                    if (line.startsWith(ABZ_M3U_LINE_PREAMBLE)) {
                        HashMap<String, String> meta = new HashMap<String, String>();
                       
                        // get track duration and title 
                        Matcher m = ABZ_M3U_LINE_PATTERN.matcher(line);
                        if (m.find()) {
                            meta.put("duration",    m.group(1));
                            meta.put("title",       m.group(2));
                        }
                        String line2 = lines[i+1].trim();
                        
                        // generate entry path
                        File w = new File(new File(playlistEntry).getParent(), line2);
                        line2 = w.getAbsolutePath().substring(1);
                        
                        // add asset
                        assets.add(new ZipAsset(line2, meta));

                        // go to the next pair of lines
                        i += 1;
                    }
                }
            }

            // close ZIP
            zipFile.close();
        } catch (Exception e) {
            // nothing
        } 
        return assets;
    } 
}

File: cordova/src_plugins/readbeyond-plugin-librarian/src/ios/FormatHandlerABZ.m
//
//            _           _            _ 
//           (_)         | |          | |
//  _ __ ___  _ _ __  ___| |_ _ __ ___| |
// | '_ ` _ \| | '_ \/ __| __| '__/ _ \ |
// | | | | | | | | | \__ \ |_| | |  __/ |
// |_| |_| |_|_|_| |_|___/\__|_|  \___|_|
//
// Author:      Alberto Pettarin (www.albertopettarin.it)
// Copyright:   Copyright 2013-2015, ReadBeyond Srl (www.readbeyond.it)
// License:     MIT
// Email:       minstrel@readbeyond.it
// Web:         http://www.readbeyond.it/minstrel/
// Status:      Production
//

#import "FormatHandlerABZ.h"
#import "RBLibrarian.h"
#import "ZipAsset.h"

@implementation FormatHandlerABZ

NSString * const ABZ_FORMAT             = @"abz";
NSString * const ABZ_DEFAULT_COVER_JPG  = @"cover.jpg";
NSString * const ABZ_DEFAULT_COVER_PNG  = @"cover.png";
NSString * const ABZ_EXTENSION_PLAYLIST = @".m3u";
NSString * const ABZ_FILE_METADATA      = @"metadata.txt";
NSString * const ABZ_M3U_HEADER         = @"#EXTM3U";
NSString * const ABZ_M3U_LINE_PATTERN   = @"#EXTINF:([0-9]+),(.*)";
NSString * const ABZ_M3U_LINE_PREAMBLE  = @"#EXTINF:";
NSMutableArray * ABZ_AUDIO_EXTENSIONS;

- (id)init
{
    if (self == [super init]) {
        [super setFormatName:ABZ_FORMAT];
    }

    // set extensions
    [self populateExtensionArray];

    return self;
}

- (void)populateExtensionArray
{
    ABZ_AUDIO_EXTENSIONS = [[NSMutableArray alloc] init];
    [ABZ_AUDIO_EXTENSIONS addObject: @".aac"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".flac"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".m4a"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".mp3"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".mp4"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".oga"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".ogg"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".wav"];
    [ABZ_AUDIO_EXTENSIONS addObject: @".webm"];
}

- (Publication *)parseFile:(NSString *)file
{
    Publication *p                   = [super parseFile:file];
    Format      *format              = [[Format alloc] init:ABZ_FORMAT];

    @try {
        BOOL foundMetadataFile  = FALSE;
        BOOL foundCover         = FALSE;
        BOOL foundPlaylist      = FALSE;
        NSString *zeCover       = nil;
        NSString *zePlaylist    = nil;
        NSMutableArray *assets  = [[NSMutableArray alloc] init];
        
        ZipFile *unzipFile  = [[ZipFile alloc] initWithFileName:file mode:ZipFileModeUnzip];
        NSArray *infos      = [unzipFile listFileInZipInfos];
        for (FileInZipInfo *info in infos) {
            NSString *name  = info.name;
            NSString *lower = [info.name lowercaseString];
            
            if (([lower hasSuffix:ABZ_DEFAULT_COVER_JPG]) || ([lower hasSuffix:ABZ_DEFAULT_COVER_PNG])) {
                zeCover = name;
            } else if ([lower hasSuffix:ABZ_EXTENSION_PLAYLIST]) {
                zePlaylist = name;
            } else if ([FormatHandlerAbstractZIP fileHasAllowedExtension:lower allowed:ABZ_AUDIO_EXTENSIONS]) {
                [p setIsValid:YES];
                [assets addObject:name];
            } else if ([lower hasSuffix:ABZ_FILE_METADATA]) {
                [p setIsValid:YES];
                foundMetadataFile = YES;
                [FormatHandlerAbstractZIP parseMetadataFile:unzipFile info:info format:format];
                if ([format getMetadatum:@"internalPathPlaylist"] != nil) {
                    foundPlaylist = YES;
                }
                if ([format getMetadatum:@"internalPathCover"] != nil) {
                    foundCover = YES;
                }
            }
        }
        
        // close ZIP
        [unzipFile close];
        
        // default cover found?
        if ((! foundCover) && (zeCover != nil)) {
            [format addMetadatum:zeCover forKey:@"internalPathCover"];
        }
        
        // default playlist found?
        if ((!foundPlaylist) && (zePlaylist != nil)) {
            [format addMetadatum:zePlaylist forKey:@"internalPathPlaylist"];
        }

        // set number of assets
        [format addMetadatum:[NSString stringWithFormat:@"%lu", (unsigned long)[assets count]] forKey:@"numberOfAssets"];
    }
    @catch (NSException *e) {
        // invalidate item, so it will not be added to library
        [p setIsValid:NO];
    }
    
    // add format
    [p addFormat:format];

    // extract cover
    [super extractCover:file format:format publicationId:[p getId]];
    
    return p;
}

- (NSString *)customAction:(NSString *)path parameters:(NSDictionary *)parameters
{
    if (parameters != nil) {
        NSString *command = [parameters objectForKey:@"command"];
        if (command != nil) {
            
            // get sorted list of assets
            if ([command isEqualToString:@"getSortedListOfAssets"]) {
                NSString *playlistEntry = [parameters objectForKey:@"internalPathPlaylist"];
                NSMutableArray *assets;
                if ((playlistEntry == nil) || ([playlistEntry isEqualToString:@""])) {
                    assets = [FormatHandlerAbstractZIP getListOfAssets:path allowed:ABZ_AUDIO_EXTENSIONS];
                } else {
                    assets = [FormatHandlerABZ parseM3UPlaylistEntry:path playlistEntry:playlistEntry];
                }
                return [RBLibrarian stringify:assets mainKey:@"assets"];
            }
        }
    }
    return @"";
}

+ (NSMutableArray *)parseM3UPlaylistEntry:(NSString *)path playlistEntry:(NSString *)playlistEntry
{
    NSMutableArray *assets  = [[NSMutableArray alloc] init];
    @try {
        ZipFile *unzipFile  = [[ZipFile alloc] initWithFileName:path mode:ZipFileModeUnzip];
        NSArray *lines      = [[FormatHandlerAbstractZIP getZipEntryText:unzipFile entryPath:playlistEntry]componentsSeparatedByString:@"\n"];
        
        // check that the first line starts with the M3U header
        if (([lines count] > 0) && ([[lines objectAtIndex:0] hasPrefix:ABZ_M3U_HEADER])) {
            
            NSRegularExpression *regex = [[NSRegularExpression alloc] initWithPattern:ABZ_M3U_LINE_PATTERN options:0 error:nil];
            
            for (int i = 1; i < [lines count]; i++) {
                NSString *line  = [[lines objectAtIndex:i] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
                
                // if line starts with the M3U preamble, parse it
                if ([line hasPrefix:ABZ_M3U_LINE_PREAMBLE]) {
                    NSMutableDictionary *meta = [[NSMutableDictionary alloc] init];
                    
                    NSRange r = NSMakeRange(0, [line length]);
                    NSArray *matches = [regex matchesInString:line options:0 range:r];
                    if ([matches count] > 0) {
                        [meta setObject:[[line substringWithRange:[[matches objectAtIndex:0] rangeAtIndex:1]] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] forKey:@"duration"];
                        [meta setObject:[[line substringWithRange:[[matches objectAtIndex:0] rangeAtIndex:2]] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] forKey:@"title"];
                    }
                    
                    // add asset
                    NSString *line2 = [[lines objectAtIndex:i+1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
                    line2 = [[playlistEntry stringByDeletingLastPathComponent] stringByAppendingPathComponent: line2];
                    
                    // add asset
                    [assets addObject:[[ZipAsset alloc] init:line2 medatada:meta]];
                    
                    // go to the next pair of lines
                    i += 1;
                }
            }
        }
        
        [unzipFile close];
    }
    @catch (NSException *e) {
        // nop
    }
    return assets;
}

@end
